### PR TITLE
StartProgram Event Handling

### PR DESCRIPTION
**Pull request by codeNinjaDev**
Builds on Event Reader Subscriber infrastructure to handle StartProgramEvents:
Based on https://github.com/cdapio/cdap/pull/15223

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: Add a newline to separate static vs non-static fields.

Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: Also group all `final` fields together, followed by non final fields.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: Use `ExecutorService` as the type instead of the concrete class.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: Not needed.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: When is the executor being shutdown? If you don't shutdown it down, it can hang the process for up to 60 seconds.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: I might missed it in another PR. Why calling this `Handler`? Generally `Handler` is for handling http requests in CDAP.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: Log with the program reference as well.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: Make it a top level class instead of inner class.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventHandlerTest.java
Comment (chtyim) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventHandlerTest.java]: Call it `LOG` to be consistent with the existing code.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: Shutting down in shutdown method now

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventHandler.java]: Changed to EventSubscriber

File: cdap-app-fabric/pom.xml
Comment (masoud-io) [inline cdap-app-fabric/pom.xml]: This change should be in the first cdap PR.

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java
Comment (masoud-io) [inline cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java]: let's make sure all these constants start with event.reader.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventReaderExtensionProvider.java
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventReaderExtensionProvider.java]: cConf.get also allows to specify default value.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventReaderExtensionProvider.java
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventReaderExtensionProvider.java]: Not needed as this will be the default behavior.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventReaderExtensionProvider.java
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventReaderExtensionProvider.java]: Here, by `enabled` , one might think that they are initialized, which is not the case.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java]: Publish to TMS is an implementation. It simply calls startProgram method on them.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java]: You should provide default value for this in cdap-default.xml

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java]: what if initialize throws an exception for one of the readers?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java]: why pull 1? It should be configurable with default value of 1.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java]: We have ProgamRunId. Let's log that instead of this approach.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java]: Should this be in CConf?

**Review by masoud-io — APPROVED** [review]
LGTM

**Review by masoud-io — APPROVED** [review]
LGTM. Couples of notes:
1. Make sure to squash commits before merge. 
2. Make sure to submit after event handling infra PR. 
3. Make sure to get a second LGTM

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java]: Formatting looks off here.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java]: nit: change the name to update handler to subscriber

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventSubscriberTest.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventSubscriberTest.java]: nit: remove extra new lines

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventSubscriberTest.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventSubscriberTest.java]: suggestion: The catch blocks here and for processEvents are not necessary, the test will fail by default if an exception occur.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventSubscriberTest.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventSubscriberTest.java]: Suggestion: Since the DummyEventReader is intended for testing, it can take events in constructor and return that on getMessages(). Mocking will not be necessary then.

**Comment by codeNinjaDev** [comment]
Tested in sandbox (without enabling event subscribing), deploying and running a pipeline is successful

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventReaderExtensionProvider.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventReaderExtensionProvider.java]: Suggestion: This message could be misleading. Change it to something like " Enabled event readers in config are "

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/dummy/DummyEventReader.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/dummy/DummyEventReader.java]: nit: A loop may be unnecessary here. ArrayList takes collection in constructor.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventSubscriberTest.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/events/StartProgramEventSubscriberTest.java]: Better to do non static initialization in setup() since this is calling a method.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java]: Is this method required to be defined here? This will force all subclasses to override it. If it is just optional behavior, the subclass could just override Object's toString().

Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java]: Forcing subscribers to override it allows logging to be a bit more standardized and consistent

Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java]: WDYT @masoud-io ?

Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java]: Greeshma has a valid point. Why not to implement toString here? One can override it in subclasses if it's needed then.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java:33]: nit: Add @override tag

**Review by greeshmaswaminathan — APPROVED** [review]

**Pull request merged**
